exchange/exchange.go

```diff
@@ -0,0 +1,223 @@
+package exchange
+
+import (
+	"github.com/mxmCherry/openrtb"
+	"github.com/prebid/prebid-server/openrtb_ext"
+	"context"
+	"time"
+	"net/http"
+	"encoding/json"
+	"fmt"
+)
+
+// Exchange runs Auctions. Implementations must be threadsafe, and will be shared across many goroutines.
+type Exchange interface {
+	// HoldAuction executes an OpenRTB v2.5 Auction.
+	HoldAuction(ctx context.Context, bidRequest *openrtb.BidRequest) (*openrtb.BidResponse, error)
+}
+
+type exchange struct {
+	// The list of adapters we will consider for this auction
+	adapters []openrtb_ext.BidderName
+	adapterMap map[openrtb_ext.BidderName]adaptedBidder
+}
+
+// Container to pass out response ext data from the GetAllBids goroutines back into the main thread
+type seatResponseExtra struct {
+	ResponseTimeMillis int
+	Errors []string
+}
+
+type bidResponseWrapper struct {
+	adapterBids *pbsOrtbSeatBid
+	adapterExtra *seatResponseExtra
+	bidder openrtb_ext.BidderName
+}
+
+func NewExchange(client *http.Client) Exchange {
+	e := new(exchange)
+
+	e.adapterMap = newAdapterMap(client)
+	e.adapters = make([]openrtb_ext.BidderName, 0, len(e.adapterMap))
+	for a, _ := range e.adapterMap {
+		e.adapters = append(e.adapters, a)
+	}
```

Comment:
I know it's late, but I believe it is slightly better to **not** use `append` when we know the size already.

Reply:
*nods*, I think that's a good point... it probably is more efficient to index into it directly. Will try to remember it in the future.